File: scripts/import-inventory.js
/**
 * Import Inventory from CSV to Umbrel PostgreSQL
 * Usage: node scripts/import-inventory.js /path/to/inventory.csv
 */

const fs = require('fs');
const { Pool } = require('pg');

// Umbrel PostgreSQL connection
const pool = new Pool({
  host: 'umbrel.local',
  port: 5432,
  database: 'aggrandize_business',
  user: 'aggrandize',
  password: 'AggrandizeDB2024',
});

// Parse price like "$1,800" or "$75" to number
function parsePrice(priceStr) {
  if (!priceStr || priceStr === '-') return null;
  const cleaned = priceStr.replace(/[$,]/g, '').trim();
  const num = parseFloat(cleaned);
  return isNaN(num) ? null : num;
}

// Parse traffic like "74K", "158.1K", "1.2M" to number
function parseTraffic(trafficStr) {
  if (!trafficStr || trafficStr === '-') return null;
  const cleaned = trafficStr.trim().toUpperCase();

  let multiplier = 1;
  let value = cleaned;

  if (cleaned.endsWith('K')) {
    multiplier = 1000;
    value = cleaned.slice(0, -1);
  } else if (cleaned.endsWith('M')) {
    multiplier = 1000000;
    value = cleaned.slice(0, -1);
  }

  const num = parseFloat(value);
  return isNaN(num) ? null : Math.round(num * multiplier);
}

// Parse boolean like "Yes", "No", "-", "Nofollow"
function parseBool(str) {
  if (!str) return false;
  const cleaned = str.trim().toLowerCase();
  return cleaned === 'yes' || cleaned === 'true' || cleaned === '1';
}

// Parse integer
function parseInt(str) {
  if (!str || str === '-') return null;
  const num = Number.parseInt(str, 10);
  return isNaN(num) ? null : num;
}

// Clean website URL - extract domain
function cleanWebsite(url) {
  if (!url) return null;
  let cleaned = url.trim();
  // Remove protocol
  cleaned = cleaned.replace(/^https?:\/\//, '');
  // Remove www.
  cleaned = cleaned.replace(/^www\./, '');
  // Remove trailing slash
  cleaned = cleaned.replace(/\/$/, '');
  return cleaned || null;
}

// Parse CSV line (handles quoted fields with commas)
function parseCSVLine(line) {
  const result = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];

    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === ',' && !inQuotes) {
      result.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }
  result.push(current.trim());

  return result;
}

async function createTableIfNotExists(client) {
  console.log('📋 Dropping and recreating website_inventory table...');

  // Drop existing table and recreate
  await client.query('DROP TABLE IF EXISTS website_inventory CASCADE');

  await client.query(`
    CREATE TABLE website_inventory (
      id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
      website VARCHAR(255) UNIQUE NOT NULL,
      contact VARCHAR(255),
      client_price DECIMAL(10,2),
      price DECIMAL(10,2),
      domain_rating INTEGER CHECK (domain_rating >= 0 AND domain_rating <= 100),
      da INTEGER CHECK (da >= 0 AND da <= 100),
      backlinks INTEGER DEFAULT 0,
      organic_traffic INTEGER DEFAULT 0,
      us_traffic INTEGER DEFAULT 0,
      uk_traffic INTEGER DEFAULT 0,
      canada_traffic INTEGER DEFAULT 0,
      is_indexed BOOLEAN DEFAULT true,
      ai_overview BOOLEAN DEFAULT false,
      chatgpt BOOLEAN DEFAULT false,
      perplexity BOOLEAN DEFAULT false,
      gemini BOOLEAN DEFAULT false,
      copilot BOOLEAN DEFAULT false,
      do_follow BOOLEAN DEFAULT true,
      news BOOLEAN DEFAULT false,
      sponsored BOOLEAN DEFAULT false,
      cbd BOOLEAN DEFAULT false,
      casino BOOLEAN DEFAULT false,
      dating BOOLEAN DEFAULT false,
      crypto BOOLEAN DEFAULT false,
      category VARCHAR(500),
      tat INTEGER DEFAULT 0,
      status VARCHAR(20) DEFAULT 'active',
      notes TEXT,
      created_by UUID,
      created_at TIMESTAMPTZ DEFAULT NOW(),
      updated_at TIMESTAMPTZ DEFAULT NOW()
    );
  `);

  console.log('✅ Table ready');
}

async function importInventory(csvPath) {
  console.log('📂 Reading CSV file:', csvPath);

  const content = fs.readFileSync(csvPath, 'utf-8');
  const lines = content.split('\n').filter(line => line.trim());

  console.log(`📊 Found ${lines.length - 1} rows (excluding header)`);

  // Parse header
  const header = parseCSVLine(lines[0]);
  console.log('📋 Columns:', header.join(', '));

  // Create column index map
  const colIndex = {};
  header.forEach((col, idx) => {
    colIndex[col.toLowerCase().trim()] = idx;
  });

  // Process rows
  const records = [];
  const errors = [];

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    if (!line.trim()) continue;

    const cols = parseCSVLine(line);

    // Skip empty rows
    const website = cleanWebsite(cols[colIndex['magazines']]);
    if (!website) {
      continue;
    }

    try {
      const record = {
        website,
        da: parseInt(cols[colIndex['da']]),
        domain_rating: parseInt(cols[colIndex['as']]),
        client_price: parsePrice(cols[colIndex['price']]),
        organic_traffic: parseTraffic(cols[colIndex['new ot']]),
        is_indexed: parseBool(cols[colIndex['indexed']]),
        category: cols[colIndex['category/section']]?.trim() || null,
        do_follow: parseBool(cols[colIndex['do follow']]),
        news: parseBool(cols[colIndex['news']]),
        sponsored: parseBool(cols[colIndex['sponsored']]),
        cbd: parseBool(cols[colIndex['cbd']]),
        casino: parseBool(cols[colIndex['casino']]),
        dating: parseBool(cols[colIndex['dating']]),
        crypto: parseBool(cols[colIndex['crypto']]),
        contact: cols[colIndex['contact']]?.trim() || null,
        tat: parseInt(cols[colIndex['tat']]),
        status: 'active',
      };

      records.push(record);
    } catch (err) {
      errors.push({ line: i + 1, error: err.message, data: line });
    }
  }

  console.log(`✅ Parsed ${records.length} valid records`);
  if (errors.length > 0) {
    console.log(`⚠️  ${errors.length} rows with errors (skipped)`);
  }

  // Insert into database
  console.log('\n🔄 Connecting to Umbrel PostgreSQL...');

  const client = await pool.connect();

  try {
    // Create table if not exists
    await createTableIfNotExists(client);

    let inserted = 0;
    let updated = 0;
    let failed = 0;

    for (const record of records) {
      try {
        // Use UPSERT (INSERT ON CONFLICT UPDATE) - each record in its own transaction
        const result = await client.query(`
          INSERT INTO website_inventory (
            website, da, domain_rating, client_price, organic_traffic,
            is_indexed, category, do_follow, news, sponsored,
            cbd, casino, dating, crypto, contact, tat, status
          ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17)
          ON CONFLICT (website) DO UPDATE SET
            da = EXCLUDED.da,
            domain_rating = EXCLUDED.domain_rating,
            client_price = EXCLUDED.client_price,
            organic_traffic = EXCLUDED.organic_traffic,
            is_indexed = EXCLUDED.is_indexed,
            category = EXCLUDED.category,
            do_follow = EXCLUDED.do_follow,
            news = EXCLUDED.news,
            sponsored = EXCLUDED.sponsored,
            cbd = EXCLUDED.cbd,
            casino = EXCLUDED.casino,
            dating = EXCLUDED.dating,
            crypto = EXCLUDED.crypto,
            contact = EXCLUDED.contact,
            tat = EXCLUDED.tat,
            updated_at = NOW()
          RETURNING (xmax = 0) AS inserted
        `, [
          record.website,
          record.da,
          record.domain_rating,
          record.client_price,
          record.organic_traffic,
          record.is_indexed,
          record.category,
          record.do_follow,
          record.news,
          record.sponsored,
          record.cbd,
          record.casino,
          record.dating,
          record.crypto,
          record.contact,
          record.tat,
          record.status,
        ]);

        if (result.rows[0].inserted) {
          inserted++;
        } else {
          updated++;
        }
      } catch (err) {
        failed++;
        console.error(`❌ Failed to insert ${record.website}:`, err.message);
      }
    }

    console.log('\n📊 Import Summary:');
    console.log(`   ✅ Inserted: ${inserted}`);
    console.log(`   🔄 Updated: ${updated}`);
    console.log(`   ❌ Failed: ${failed}`);
    console.log(`   📦 Total: ${inserted + updated + failed}`);

    // Verify count
    const countResult = await client.query('SELECT COUNT(*) FROM website_inventory');
    console.log(`\n📈 Total records in database: ${countResult.rows[0].count}`);

  } catch (err) {
    throw err;
  } finally {
    client.release();
  }

  await pool.end();
  console.log('\n✅ Import complete!');
}

// Run import
const csvPath = process.argv[2] || '/Users/dhanapale/Downloads/inventory.csv';
importInventory(csvPath).catch(err => {
  console.error('❌ Import failed:', err);
  process.exit(1);
});

File: scripts/import_inventory.py
#!/usr/bin/env python3
"""
Import inventory.csv to website_inventory table via Umbrel API
"""

import csv
import json
import requests
import re
import sys

API_URL = "https://api.aggrandizedigital.com/query"
API_KEY = "e622c42ee210f3ad0af8ec91ec92d164cfb16e3e9ecec5e991eb6fe2ece2180e"
ADMIN_KEY = "v4+qGvUcO3A0Gqhstn84pWu9v2E2xnZ5DlPCtvafvKs="

def parse_price(price_str):
    """Parse price like '$75' or '$1,800' to decimal"""
    if not price_str or price_str.strip() == '':
        return 0
    # Remove $ and commas
    cleaned = price_str.replace('$', '').replace(',', '').strip()
    try:
        return float(cleaned)
    except:
        return 0

def parse_traffic(traffic_str):
    """Parse traffic like '74K' or '158.1K' to integer"""
    if not traffic_str or traffic_str.strip() == '':
        return 0
    cleaned = traffic_str.upper().replace(',', '').strip()
    try:
        if 'K' in cleaned:
            return int(float(cleaned.replace('K', '')) * 1000)
        elif 'M' in cleaned:
            return int(float(cleaned.replace('M', '')) * 1000000)
        else:
            return int(float(cleaned))
    except:
        return 0

def parse_bool(val):
    """Parse Yes/No/- to boolean"""
    if not val:
        return False
    val = val.strip().lower()
    return val in ['yes', 'true', '1']

def parse_dofollow(val):
    """Parse Do Follow field (Yes/No/Nofollow/-)"""
    if not val:
        return True  # Default to dofollow
    val = val.strip().lower()
    if val in ['yes', 'true', '1']:
        return True
    if val in ['no', 'nofollow', 'false', '0', '-']:
        return False
    return True

def clean_website(url):
    """Clean website URL to just domain"""
    if not url:
        return None
    # Remove protocol
    url = re.sub(r'^https?://', '', url)
    # Remove www.
    url = re.sub(r'^www\.', '', url)
    # Remove trailing slash
    url = url.rstrip('/')
    return url

def parse_int(val):
    """Parse integer safely"""
    if not val or val.strip() == '':
        return None
    try:
        return int(val)
    except:
        return None

def insert_website(row):
    """Insert a single website into database"""
    website = clean_website(row.get('Magazines', ''))
    if not website:
        return None, "No website URL"

    # Build data
    data = {
        'website': website,
        'contact': row.get('Contact', '').strip() or None,
        'category': row.get('Category/Section', '').strip() or None,
        'da': parse_int(row.get('DA', '')),
        'domain_rating': parse_int(row.get('AS', '')),
        'organic_traffic': parse_traffic(row.get('New OT', '')),
        'client_price': parse_price(row.get('Price', '')),
        'is_indexed': parse_bool(row.get('Indexed', '')),
        'do_follow': parse_dofollow(row.get('Do Follow', '')),
        'news': parse_bool(row.get('News', '')),
        'sponsored': parse_bool(row.get('Sponsored', '')),
        'cbd': parse_bool(row.get('CBD', '')),
        'casino': parse_bool(row.get('Casino', '')),
        'dating': parse_bool(row.get('Dating', '')),
        'crypto': parse_bool(row.get('Crypto', '')),
        'tat': parse_int(row.get('TAT', '')),
        'status': 'active'
    }

    # Build SQL
    fields = [k for k, v in data.items() if v is not None]
    values = [v for v in data.values() if v is not None]
    placeholders = [f'${i+1}' for i in range(len(fields))]

    sql = f"INSERT INTO website_inventory ({', '.join(fields)}) VALUES ({', '.join(placeholders)}) ON CONFLICT (website) DO NOTHING RETURNING id, website"

    try:
        response = requests.post(
            API_URL,
            headers={
                'Content-Type': 'application/json',
                'X-API-Key': API_KEY,
                'X-ADMIN-KEY': ADMIN_KEY
            },
            json={'sql': sql, 'params': values},
            timeout=30
        )
        result = response.json()
        if result.get('success'):
            if result.get('rowCount', 0) > 0:
                return result['rows'][0], None
            else:
                return None, "Duplicate (skipped)"
        else:
            return None, result.get('error', 'Unknown error')
    except Exception as e:
        return None, str(e)

def main():
    import time

    csv_path = '/Users/dhanapale/Downloads/inventory.csv'
    start_from = int(sys.argv[1]) if len(sys.argv) > 1 else 0

    print(f"Reading {csv_path}...")

    success_count = 0
    skip_count = 0
    error_count = 0
    errors = []

    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        rows = list(reader)

    print(f"Found {len(rows)} rows to import (starting from row {start_from})")

    for i, row in enumerate(rows):
        if i < start_from:
            continue

        # Add delay to avoid rate limiting (100 req/60s = ~1.7 req/s)
        time.sleep(0.7)
        # Skip empty rows
        if not row.get('Magazines') or row.get('Magazines', '').strip() == '':
            skip_count += 1
            continue

        # Retry logic for rate limiting
        retries = 3
        result = None
        error = None
        for attempt in range(retries):
            result, error = insert_website(row)
            if error and 'Too many requests' in error:
                print(f"  Rate limited at row {i+1}, waiting 60s... (attempt {attempt+1}/{retries})")
                time.sleep(60)
                continue
            break

        if result:
            success_count += 1
            if success_count % 50 == 0:
                print(f"  Imported {success_count} websites...")
        elif error == "Duplicate (skipped)":
            skip_count += 1
        else:
            error_count += 1
            errors.append(f"Row {i+1} ({row.get('Magazines', 'unknown')}): {error}")

    print(f"\n=== Import Complete ===")
    print(f"Successfully imported: {success_count}")
    print(f"Skipped (duplicates/empty): {skip_count}")
    print(f"Errors: {error_count}")

    if errors and error_count <= 10:
        print("\nErrors:")
        for e in errors:
            print(f"  - {e}")
    elif errors:
        print(f"\nFirst 10 errors:")
        for e in errors[:10]:
            print(f"  - {e}")

if __name__ == '__main__':
    main()
